prow/plugins/wip/wip-label.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package wip will label a PR a work-in-progress if the author provides
+// a prefix to their pull request title to the same effect. The submit-
+// queue will not merge pull requests with the work-in-progress label.
+// The label will be removed when the title changes to no longer begin
+// with the prefix.
+package wip
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/Sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const (
+	label      = "do-not-merge/work-in-progress"
+	pluginName = "wip"
+)
+
+var (
+	titleRegex = regexp.MustCompile(`(?i)^(\[WIP\]|WIP)`)
+)
+
+type event struct {
+	org         string
+	repo        string
+	number      int
+	hasLabel    bool
+	needsLabel  bool
+	commentBody string
+}
+
+func init() {
+	plugins.RegisterPullRequestHandler(pluginName, handlePullRequest)
+}
+
+// Strict subset of *github.Client methods.
+type githubClient interface {
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+	AddLabel(owner, repo string, number int, label string) error
+	RemoveLabel(owner, repo string, number int, label string) error
+	CreateComment(org, repo string, number int, comment string) error
+	DeleteComment(org, repo string, ID int) error
+}
+
+func handlePullRequest(pc plugins.PluginClient, pe github.PullRequestEvent) error {
+	// These are the only actions indicating the PR title may have changed.
+	if pe.Action != github.PullRequestActionOpened && pe.Action != github.PullRequestActionEdited && pe.Action != github.PullRequestActionSynchronize {
+		return nil
+	}
+
+	var (
+		org    = pe.PullRequest.Base.Repo.Owner.Login
+		repo   = pe.PullRequest.Base.Repo.Name
+		number = pe.PullRequest.Number
+		author = pe.PullRequest.User.Login
+		title  = pe.PullRequest.Title
+	)
+
+	currentLabels, err := pc.GitHubClient.GetIssueLabels(org, repo, number)
```

Comment:
Does the opened event already include the labels? If so can we just extract them from `pe`?

Reply:
So the struct contains this:

```go
// PullRequestEvent is what GitHub sends us when a PR is changed.
type PullRequestEvent struct {
	Label       Label                  `json:"label"`
}
```

However that's a singular label and the [API doc](https://developer.github.com/v3/activity/events/types/#pullrequestevent) doesn't list a top-level `"label"` field so I'm not really sure what that is or where it comes from. I think we need to do the API call here.